File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/errortracking/BugsnagProvider.java
package com.aoneahsan.unifiedtracking.providers.errortracking;

import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.aoneahsan.unifiedtracking.providers.ErrorTrackingProvider;
import com.bugsnag.android.Bugsnag;
import com.bugsnag.android.Configuration;

import java.util.Map;

/**
 * Bugsnag error-tracking provider for Android.
 *
 * Uses {@code com.bugsnag.android.Bugsnag} (the Bugsnag Android SDK, artifact
 * {@code com.bugsnag:bugsnag-android}). The SDK is initialized with a
 * {@code Configuration} object via {@code Bugsnag.start(context, cfg)}.
 *
 * NOTE(unverified): built against Bugsnag Android SDK 6.x API. Confirm exact
 * signatures (Configuration constructor, setReleaseStage, Bugsnag.notify with
 * OnErrorCallback lambda, Bugsnag.setUser, Bugsnag.pauseSession) in Android
 * Studio against the resolved SDK version.
 */
public class BugsnagProvider implements ErrorTrackingProvider {
    private static final String TAG = "Bugsnag";
    private final Context context;
    private final JSObject config;
    private boolean enabled;
    private boolean initialized;
    private boolean debugMode = false;
    // Track previous consent state so we know when to pause an active session.
    private boolean previousConsent = true;

    public BugsnagProvider(Context context, JSObject config) {
        this.context = context;
        this.config = config;
        this.enabled = config.getBoolean("enabled", true);
        this.previousConsent = this.enabled;
        this.initialized = false;

        try {
            String apiKey = config.getString("apiKey");
            if (apiKey == null || apiKey.trim().isEmpty()) {
                Log.e(TAG, "Bugsnag apiKey is missing; provider will be inert.");
                return;
            }

            // NOTE(unverified): Configuration(String) is the documented constructor in
            // Bugsnag Android 6.x. Older SDKs used Configuration.load(context). Confirm
            // against the resolved SDK.
            Configuration cfg = new Configuration(apiKey);

            String releaseStage = config.getString("releaseStage", null);
            if (releaseStage != null && !releaseStage.trim().isEmpty()) {
                cfg.setReleaseStage(releaseStage);
            }

            String appVersion = config.getString("appVersion", null);
            if (appVersion != null && !appVersion.trim().isEmpty()) {
                // NOTE(unverified): Configuration.setAppVersion(String) is present in
                // Bugsnag Android 6.x. Confirm against the resolved SDK.
                cfg.setAppVersion(appVersion);
            }

            Bugsnag.start(context, cfg);
            this.initialized = true;

            if (debugMode) {
                Log.d(TAG, "Bugsnag initialized (releaseStage=" + releaseStage + ")");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Bugsnag", e);
            this.initialized = false;
        }
    }

    @Override
    public String getName() {
        return "bugsnag";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void logError(Exception error, Map<String, Object> errorContext) {
        if (!enabled || !initialized || error == null) return;

        try {
            // NOTE(unverified): Bugsnag.notify(Throwable, OnErrorCallback) is the
            // documented Java overload in SDK 6.x. The lambda receives an Event and
            // returns boolean (true to send, false to drop). We attach the context
            // map as metadata under the "context" section.
            if (errorContext != null && !errorContext.isEmpty()) {
                final Map<String, Object> finalContext = errorContext;
                Bugsnag.notify(error, event -> {
                    try {
                        for (Map.Entry<String, Object> entry : finalContext.entrySet()) {
                            if (entry.getKey() == null || entry.getValue() == null) continue;
                            // NOTE(unverified): Event.addMetadata(section, key, value)
                            // is the documented metadata API in SDK 6.x. Confirm
                            // against the resolved SDK; some older builds used
                            // addToTab(...).
                            event.addMetadata("context", entry.getKey(), entry.getValue());
                        }
                    } catch (Throwable ignored) {
                        // Best-effort metadata attachment; never block delivery on it.
                    }
                    return true;
                });
            } else {
                Bugsnag.notify(error);
            }

            if (debugMode) {
                Log.d(TAG, "Notified Bugsnag: " + error.getMessage());
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to notify Bugsnag", e);
        }
    }

    @Override
    public void setUserContext(String userId, Map<String, Object> properties) {
        if (!enabled || !initialized) return;

        try {
            String email = null;
            String name = null;
            if (properties != null) {
                Object e = properties.get("email");
                if (e != null) email = String.valueOf(e);
                Object n = properties.get("name");
                if (n != null) name = String.valueOf(n);
            }

            // NOTE(unverified): Bugsnag.setUser(String id, String email, String name)
            // is the documented signature in SDK 6.x. Confirm against the resolved SDK.
            Bugsnag.setUser(userId, email, name);

            if (debugMode) {
                Log.d(TAG, "Set Bugsnag user: " + userId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set user context", e);
        }
    }

    @Override
    public void setConsent(boolean granted) {
        if (!initialized) return;

        try {
            // Bugsnag does not expose a runtime opt-out; we gate dispatch locally
            // via the `enabled` flag (logError/setUserContext early-return when false).
            // When transitioning from granted -> revoked, pause the active session so
            // the SDK stops auto-reporting session counts.
            if (previousConsent && !granted) {
                try {
                    // NOTE(unverified): Bugsnag.pauseSession() is the documented runtime
                    // session toggle in SDK 6.x; resumeSession() resumes it. Some older
                    // SDKs only expose autoTrackSessions in Configuration.
                    Bugsnag.pauseSession();
                } catch (Throwable ignored) {
                    // Optional best-effort; consent gate still works via the flag below.
                }
            } else if (!previousConsent && granted) {
                try {
                    Bugsnag.resumeSession();
                } catch (Throwable ignored) {
                    // Optional best-effort.
                }
            }

            this.previousConsent = granted;
            this.enabled = granted;

            if (debugMode) {
                Log.d(TAG, "Set consent: " + granted);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set consent", e);
        }
    }

    @Override
    public void reset() {
        if (!initialized) return;

        try {
            // NOTE(unverified): Bugsnag.setUser(null, null, null) clears the user
            // record in SDK 6.x. Confirm the SDK accepts null for all three args.
            Bugsnag.setUser(null, null, null);

            if (debugMode) {
                Log.d(TAG, "Reset Bugsnag user context");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset", e);
        }
    }

    @Override
    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        if (enabled) {
            Log.d(TAG, "Debug mode enabled for Bugsnag provider");
        }
    }
}

File: ios/Sources/UnifiedTrackingPlugin/ErrorTracking/BugsnagProvider.swift
import Foundation
import Bugsnag

/// Bugsnag (bugsnag-cocoa) bridge for the unified-tracking plugin.
///
/// Reads `apiKey` (required) and optional `releaseStage` from the provider
/// config and starts the Bugsnag SDK. If no API key is supplied the SDK is not
/// started and the provider stays uninitialized.
///
/// NOTE: Bugsnag does NOT expose a first-class consent toggle. When consent is
/// revoked we gate `logError` at the provider level (`consentGranted`) and
/// pause the current session. The native SDK keeps running but no events leave
/// this provider while consent is denied — documented limitation.
class BugsnagProvider: ErrorTrackingProvider {
    let name = "bugsnag"
    private(set) var isEnabled: Bool
    private(set) var isInitialized: Bool

    private let config: [String: Any]
    private var debugMode = false
    /// Tracks the consent state. Bugsnag has no opt-out toggle, so this provider
    /// gates `logError` itself when consent is revoked.
    private var consentGranted = true

    init(config: [String: Any]) {
        self.config = config
        self.isEnabled = config["enabled"] as? Bool ?? true
        self.isInitialized = false

        guard let apiKey = config["apiKey"] as? String, !apiKey.isEmpty else {
            // No API key -> cannot start Bugsnag. Leave uninitialized.
            print("[BugsnagProvider] No apiKey provided; Bugsnag not started.")
            return
        }

        // NOTE(unverified): build with Xcode — confirm `BugsnagConfiguration(_:)`
        // initializer takes the apiKey as the first positional argument and that
        // `releaseStage` / `appVersion` are writable properties in bugsnag-cocoa ~6.30.2.
        let configuration = BugsnagConfiguration(apiKey)
        if let releaseStage = config["releaseStage"] as? String {
            configuration.releaseStage = releaseStage
        }
        if let appVersion = config["appVersion"] as? String {
            configuration.appVersion = appVersion
        }

        // NOTE(unverified): build with Xcode — confirm `Bugsnag.start(with:)` class
        // method exists in bugsnag-cocoa ~6.30.2 (older SDKs used `Bugsnag.start()`
        // with the config supplied via `Bugsnag.startBugsnagWithApiKey(...)`).
        Bugsnag.start(with: configuration)

        self.isInitialized = true
    }

    func logError(_ error: Error, context: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted else { return }
        // NOTE(unverified): build with Xcode — confirm
        // `Bugsnag.notify(_:block:)` with a `BugsnagEvent` callback returning Bool,
        // and `event.addMetadata(_:key:section:)` argument labels in bugsnag-cocoa ~6.30.2.
        Bugsnag.notify(error) { event in
            for (key, value) in context {
                event.addMetadata(value, key: key, section: "context")
            }
            return true
        }
    }

    func setUserContext(_ userId: String, properties: [String: Any]) {
        guard isEnabled && isInitialized else { return }
        // NOTE(unverified): build with Xcode — confirm
        // `Bugsnag.setUser(_:withEmail:andName:)` signature in bugsnag-cocoa ~6.30.2.
        let email = properties["email"] as? String
        let name = properties["name"] as? String
        Bugsnag.setUser(userId, withEmail: email, andName: name)
    }

    func setConsent(_ granted: Bool) {
        consentGranted = granted
        // NOTE: Bugsnag has no built-in opt-out toggle, so we pause the current
        // session when consent is revoked. Future `logError` calls are gated by
        // `consentGranted` inside this provider.
        // NOTE(unverified): build with Xcode — confirm `Bugsnag.pauseSession()` and
        // `Bugsnag.resumeSession()` class methods exist in bugsnag-cocoa ~6.30.2.
        if granted {
            Bugsnag.resumeSession()
        } else {
            Bugsnag.pauseSession()
        }
    }

    func reset() {
        // Clear the user attribution on the running Bugsnag instance.
        Bugsnag.setUser(nil, withEmail: nil, andName: nil)
    }

    func setDebugMode(_ enabled: Bool) {
        // Bugsnag's verbose logging is fixed at SDK start
        // (`BugsnagConfiguration.enabledBreadcrumbTypes`, etc.). We retain the
        // flag here for parity; flipping it at runtime is not supported by the SDK.
        debugMode = enabled
    }
}
